contracts/xastro_token/src/contract.rs

```diff
@@ -0,0 +1,1308 @@
+        }
+        QueryMsg::TotalSupplyAt { block } => to_binary(&get_total_supply_at(deps.storage, block)?),
+        QueryMsg::TokenInfo {} => to_binary(&query_token_info(deps)?),
+        QueryMsg::Minter {} => to_binary(&query_minter(deps)?),
+        QueryMsg::Allowance { owner, spender } => {
+            to_binary(&query_allowance(deps, owner, spender)?)
+        }
+        QueryMsg::AllAllowances {
+            owner,
+            start_after,
+            limit,
+        } => to_binary(&query_all_allowances(deps, owner, start_after, limit)?),
+        QueryMsg::AllAccounts { start_after, limit } => {
+            to_binary(&query_all_accounts(deps, start_after, limit)?)
+        }
+        QueryMsg::MarketingInfo {} => to_binary(&query_marketing_info(deps)?),
+        QueryMsg::DownloadLogo {} => to_binary(&query_download_logo(deps)?),
+    }
+}
+
+/// ## Description
+/// Returns an [`StdError`] on failure, otherwise returns balance of the given address.
+/// ## Params
+/// * **deps** is the object of type [`Deps`].
+///
+/// * **address** is the object of type [`String`].
+pub fn query_balance(deps: Deps, address: String) -> StdResult<BalanceResponse> {
+    let address = deps.api.addr_validate(&address)?;
+    let balance = BALANCES
+        .may_load(deps.storage, &address)?
+        .unwrap_or_default();
+    Ok(BalanceResponse { balance })
+}
+
+/// ## Description
+/// Returns an [`StdError`] on failure, otherwise returns balance of the given address at the given block.
+/// ## Params
+/// * **deps** is the object of type [`Deps`].
+///
+/// * **address** is the object of type [`String`].
+///
+/// * **block** is the object of type [`u64`].
+pub fn query_balance_at(deps: Deps, address: String, block: u64) -> StdResult<BalanceResponse> {
+    let address = deps.api.addr_validate(&address)?;
+    let balance = BALANCES
+        .may_load_at_height(deps.storage, &address, block)?
+        .unwrap_or_default();
+    Ok(BalanceResponse { balance })
+}
+
+/// ## Description
+/// Returns an [`StdError`] on failure, otherwise returns balance of the given address at the given block.
+/// ## Params
+/// * **deps** is the object of type [`Deps`].
+///
+/// * **start_after** is an [`Option`] field object of type [`String`].
+///
+/// * **limit** is an [`Option`] field object of type [`u32`].
+pub fn query_all_accounts(
+    deps: Deps,
+    start_after: Option<String>,
+    limit: Option<u32>,
+) -> StdResult<AllAccountsResponse> {
+    let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+    let start = start_after.map(Bound::exclusive);
+
+    let accounts: Result<Vec<_>, _> = BALANCES
+        .range(deps.storage, start, None, Order::Ascending)
+        .take(limit)
+        .map(StdResult::unwrap)
+        .map(|(a, _)| String::from_utf8(a))
+        .collect();
+
+    Ok(AllAccountsResponse {
+        accounts: accounts?,
+    })
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Let's move tests to separate file, as we do in other contracts

Reply:
Done.